File: api/v1/endpoints/favorites.py
# -*- coding: utf-8 -*-
"""System-wide favorites endpoints (anonymous user shared)."""

from __future__ import annotations

from fastapi import APIRouter, Response
from pydantic import BaseModel, Field

from src.storage import DatabaseManager

router = APIRouter()


class FavoriteRequest(BaseModel):
    """Add favorite request body."""

    model_config = {"populate_by_name": True}

    code: str = Field(..., description="Stock or ETF code")
    type: str = Field(..., alias="type", description="'stock' or 'etf'")


class FavoritesResponse(BaseModel):
    """Favorites list response."""

    stocks: list[str] = Field(default_factory=list)
    etfs: list[str] = Field(default_factory=list)


@router.get(
    "",
    response_model=FavoritesResponse,
    summary="Get all favorites",
)
def get_favorites() -> FavoritesResponse:
    """Return system-wide favorites (shared by all anonymous users)."""
    db = DatabaseManager.get_instance()
    return FavoritesResponse(
        stocks=db.get_favorites("stock"),
        etfs=db.get_favorites("etf"),
    )


@router.post(
    "",
    status_code=201,
    summary="Add a favorite",
)
def add_favorite(body: FavoriteRequest) -> dict:
    """Add a stock or ETF to system-wide favorites."""
    db = DatabaseManager.get_instance()
    db.add_favorite(body.code, body.type)
    return {"ok": True}


@router.delete(
    "/{code}",
    summary="Remove a favorite",
)
def remove_favorite(code: str, type: str = "stock") -> Response:
    """Remove a stock or ETF from system-wide favorites."""
    db = DatabaseManager.get_instance()
    db.remove_favorite(code, type)
    return Response(status_code=204)

File: apps/dsa-web/src/api/favorites.ts
import apiClient from './index';

const DEVICE_ID_KEY = 'dsa_device_id';

function getDeviceId(): string {
  let id = localStorage.getItem(DEVICE_ID_KEY);
  if (!id) {
    id = crypto.randomUUID();
    localStorage.setItem(DEVICE_ID_KEY, id);
  }
  return id;
}

function favoritesHeaders() {
  return { 'X-Device-ID': getDeviceId() };
}

export interface FavoritesResponse {
  stocks: string[];
  etfs: string[];
}

export interface AddFavoriteRequest {
  code: string;
  type: 'stock' | 'etf';
}

export async function getFavorites(): Promise<FavoritesResponse> {
  const { data } = await apiClient.get<FavoritesResponse>('/api/v1/favorites', {
    headers: favoritesHeaders(),
  });
  return data;
}

export async function addFavorite(code: string, type: 'stock' | 'etf'): Promise<void> {
  await apiClient.post<void>(
    '/api/v1/favorites',
    { code, type } as AddFavoriteRequest,
    { headers: favoritesHeaders() }
  );
}

export async function removeFavorite(code: string, type: 'stock' | 'etf' = 'stock'): Promise<void> {
  await apiClient.delete(`/api/v1/favorites/${code}?type=${type}`, {
    headers: favoritesHeaders(),
  });
}